crates/ty_python_semantic/src/types/signatures.rs

```diff
@@ -348,6 +377,114 @@ impl<'db> CallableSignature<'db> {
         relation_visitor: &HasRelationToVisitor<'db>,
         disjointness_visitor: &IsDisjointVisitor<'db>,
     ) -> ConstraintSet<'db> {
+        if relation.is_constraint_set_assignability() {
+            // TODO: Oof, maybe ParamSpec needs to live at CallableSignature, not Signature?
+            let self_is_single_paramspec = Self::signatures_is_single_paramspec(self_signatures);
+            let other_is_single_paramspec = Self::signatures_is_single_paramspec(other_signatures);
+
+            // If either callable is a ParamSpec, the constraint set should bind the ParamSpec to
+            // the other callable's signature. We also need to compare the return types — for
+            // instance, to verify in `Callable[P, int]` that the return type is assignable to
+            // `int`, or in `Callable[P, T]` to bind `T` to the return type of the other callable.
+            match (self_is_single_paramspec, other_is_single_paramspec) {
```

Comment:
Would this need to be moved to `Signature` once overloads are supported?

During the initial implementation of `ParamSpec`, when I started off by trying to use the new solver, I had a similar `match` expression in `Signature::has_relation_impl_inner` after the `is_gradual` checks where I'd try to add similar constraints.